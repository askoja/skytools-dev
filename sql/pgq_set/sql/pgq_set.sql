
\set ECHO none
\i ../txid/txid.sql
\i ../pgq/pgq.sql
\i structure/pgq_set.sql
\i structure/functions.sql
\set ECHO all

select * from pgq_set.add_member('aset', 'node1', 'dbname=node1', false);
select * from pgq_set.add_member('aset', 'node2', 'dbname=node2', false);
select * from pgq_set.add_member('aset', 'node3', 'dbname=node3', false);
select * from pgq_set.add_member('aset', 'node4', 'dbname=node4', false);
select * from pgq_set.get_member_info('aset');

select * from pgq_set.remove_member('aset', 'node4');
select * from pgq_set.get_member_info('aset');

select * from pgq_set.create_node('aset', 'root', 'node1', 'worker', null, null, null);

select * from pgq_set.get_node_info('aset');

select * from pgq_set.is_root('q');
select * from pgq_set.is_root('aset');
select * from pgq_set.is_root(null);

