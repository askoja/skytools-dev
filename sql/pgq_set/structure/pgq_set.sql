
create schema pgq_set;
grant usage on schema pgq_set to public;

-- ----------------------------------------------------------------------
-- Table: pgq_set.member_info
--
--      Static table that just lists all members in set.
--
-- Columns:
--      set_name        - set name
--      node_name       - node name
--      node_location   - libpq connect string for connecting to node
--      online          - whether the node is available
-- ----------------------------------------------------------------------
create table pgq_set.member_info (
    set_name        text not null,
    node_name       text not null,
    node_location   text not null,
    dead            boolean not null default false,

    primary key (set_name, node_name)
);

-- ----------------------------------------------------------------------
-- Table: pgq_set.local_node
--
--      Local node info.
--
-- Columns:
--      set_name            - set name
--      node_type           - local node type
--      node_name           - local node name
--      queue_name          - local queue name for set, NULL on leaf
--      provider_node       - provider node name
--      combined_set        - on 'merge-leaf' the target combined set name
--      global_watermark    - set's global watermark, set by root node
--      paused              - true if worker for this node should sleep
--      resync              - true if worker for this node needs to re-register itself on provider queue
--      up_to_date          - true if worker for this node has seen table changes
--
-- Node types:
--      root            - data + batches is generated here
--      branch          - replicates full queue contents and maybe contains some tables
--      leaf            - does not replicate queue
--      combined-root   - data from several partitions is merged here
--      combined-branch - can take over the role of combined-root
--      merge-leaf      - this node in part set is linked to combined-root/branch
-- ----------------------------------------------------------------------
create table pgq_set.set_info (
    set_name        text not null primary key,
    node_type       text not null,
    node_name       text not null,
    queue_name      text,
    provider_node   text,
    combined_set    text,

    global_watermark bigint not null,

    paused          boolean not null default false,
    resync          boolean not null default false,
    up_to_date      boolean not null default false,

    foreign key (set_name, node_name) references pgq_set.member_info,
    foreign key (set_name, provider_node) references pgq_set.member_info,
    check (node_type in ('root', 'branch', 'leaf', 'combined-root', 'combined-branch', 'merge-leaf')),
    check (case when node_type = 'root'              then (queue_name is not null and provider_node is null     and combined_set is null)
                when node_type = 'branch'            then (queue_name is not null and provider_node is not null and combined_set is null)
                when node_type = 'leaf'              then (queue_name is null     and provider_node is not null and combined_set is null)
                when node_type = 'combined-root'     then (queue_name is not null and provider_node is null     and combined_set is null)
                when node_type = 'combined-branch'   then (queue_name is not null and provider_node is not null and combined_set is null)
                when node_type = 'merge-leaf'        then (queue_name is null     and provider_node is not null and combined_set is not null)
                else false end)
);

-- ----------------------------------------------------------------------
-- Table: pgq_set.subscriber_info
--
--      Contains subscribers for a set.
--
-- Columns:
--      set_name        - set's name
--      node_name       - node name
--      worker_name     - consumer_name for node
--      local_watermark - watermark for node and it's subscribers
-- ----------------------------------------------------------------------
create table pgq_set.subscriber_info (
    set_name        text not null,
    node_name       text not null,
    worker_name     text not null,
    local_watermark bigint not null,

    primary key (set_name, node_name),
    foreign key (set_name, node_name) references pgq_set.member_info
);

-- ----------------------------------------------------------------------
-- Table: pgq_set.completed_tick
--
--      Contains completed tick_id from provider.
--
-- Columns:
--      set_name - set's name
--      tick_id  - last committed tick id
-- ----------------------------------------------------------------------
create table pgq_set.completed_tick (
    set_name        text not null primary key,
    tick_id         bigint not null,

    foreign key (set_name) references pgq_set.set_info
);

