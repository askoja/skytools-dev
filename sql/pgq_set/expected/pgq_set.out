\set ECHO none
select * from pgq_set.add_member('aset', 'node1', 'dbname=node1', false);
 ret_code | ret_note 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.add_member('aset', 'node2', 'dbname=node2', false);
 ret_code | ret_note 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.add_member('aset', 'node3', 'dbname=node3', false);
 ret_code | ret_note 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.add_member('aset', 'node4', 'dbname=node4', false);
 ret_code | ret_note 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.get_member_info('aset');
 node_name | node_location | dead 
-----------+---------------+------
 node1     | dbname=node1  | f
 node2     | dbname=node2  | f
 node3     | dbname=node3  | f
 node4     | dbname=node4  | f
(4 rows)

select * from pgq_set.remove_member('aset', 'node4');
 ret_code | ret_note 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.get_member_info('aset');
 node_name | node_location | dead 
-----------+---------------+------
 node1     | dbname=node1  | f
 node2     | dbname=node2  | f
 node3     | dbname=node3  | f
(3 rows)

select * from pgq_set.create_node('aset', 'root', 'node1', 'worker', null, null, null);
 ret_code | ret_desc 
----------+----------
      200 | Ok
(1 row)

select * from pgq_set.get_node_info('aset');
 node_type | node_name | queue_name | global_watermark | local_watermark | completed_tick | worker_name | provider_node | provider_location | paused | resync | up_to_date | combined_set | combined_type | combined_queue 
-----------+-----------+------------+------------------+-----------------+----------------+-------------+---------------+-------------------+--------+--------+------------+--------------+---------------+----------------
 root      | node1     | aset       |                1 |               1 |                | worker      |               |                   | f      | f      | f          |              |               | 
(1 row)

select * from pgq_set.is_root('q');
ERROR:  set does not exist: q
select * from pgq_set.is_root('aset');
 is_root 
---------
 t
(1 row)

select * from pgq_set.is_root(null);
ERROR:  set does not exist: <NULL>
